### ISSUE TITLE
Unable to manage file: string indices must be integers, not str

### ISSUE THREAD

**Issue by yermulnik** [saltstack/salt#26090]
`archive.extracted` started to throw an error after upgrade to 2015.5.3:
```yaml
{{ extract_path }}:
  archive.extracted:
    - source: {{ archive_file }}
    - source_hash: salt://deploy/node_{{ cluster }}/worker_oe/bin/{{ oe_file }}.md5
    - archive_format: zip
```
```yaml
(jail) [root#jail_salt][w:0][j:0][.../salt/states]> salt jail_salt state.sls_id /usr/local/etc/salt/states/deploy/node_srvt/worker_oe/bin/1.6.0.3610 master
jail_salt:
----------
          ID: /usr/local/etc/salt/states/deploy/node_srvt/worker_oe/bin/1.6.0.3610
    Function: archive.extracted
      Result: False
     Comment: Unable to manage file: string indices must be integers, not str
     Started: 07:32:15.124710
    Duration: 2173.436 ms
     Changes:

Summary
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
```

**Comment by jfindlay** [comment]
@yermulnik, thanks for the report.

**Comment by vr-jack** [comment]
@yermulnik can you try running it from the minion with salt-call -l debug ? It provides more information including your rendered yaml. Also, please confirm that your master and minion are both 2015.5.3. I was unable to duplicate this.

**Comment by yermulnik** [comment]
Salt master and minion are both on the same host:
```
(jail) [root#jail_salt][w:0][j:0][.../salt/states]> salt --version
salt 2015.5.3 (Lithium)
(jail) [root#jail_salt][w:0][j:0][.../salt/states]> salt-call --version
salt-call 2015.5.3 (Lithium)
```
state file:
```yaml
/tmp/yz/oe/oe-1.6.0.3652:
  archive.extracted:
    - source: /tmp/yz/oe/oe-1.6.0.3652.war
    - source_hash: /tmp/yz/oe/oe-1.6.0.3652.war.md5
    - archive_format: zip
```
```shell
(jail) [root#jail_salt][w:0][j:0][.../salt/states]> salt-call -l debug state.sls yz_test
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Using cached minion ID from /usr/local/etc/salt/minion_id: jail_salt
[DEBUG   ] Configuration file path: /usr/local/etc/salt/minion
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Initializing new SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded state.sls
[DEBUG   ] LazyLoaded saltutil.is_running
[DEBUG   ] LazyLoaded grains.get
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://yz_test.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'yz_test.sls'
[DEBUG   ] LazyLoaded cmd.run
[DEBUG   ] Jinja search path: ['/var/cache/salt/minion/files/base']
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/yz_test.sls:
/tmp/yz/oe/oe-1.6.0.3652:
  archive.extracted:
    - source: /tmp/yz/oe/oe-1.6.0.3652.war
    - source_hash: /tmp/yz/oe/oe-1.6.0.3652.war.md5
    - archive_format: zip

[DEBUG   ] LazyLoaded config.get
[DEBUG   ] Results of YAML rendering:
OrderedDict([('/tmp/yz/oe/oe-1.6.0.3652', OrderedDict([('archive.extracted', [OrderedDict([('source', '/tmp/yz/oe/oe-1.6.0.3652.war')]), OrderedDict([('source_hash', '/tmp/yz/oe/oe-1.6.0.3652.war.md5')]), OrderedDict([('archive_format', 'zip')])])]))])
[DEBUG   ] Error loading module.ipmi: No module named pyghmi.ipmi
[DEBUG   ] Error loading module.npm: npm execution module could not be loaded because the npm binary could not be located
[DEBUG   ] Error loading module.tls: ['PyOpenSSL version 0.14 or later must be installed before  this module can be used.']
[DEBUG   ] Error loading module.nacl: libnacl import error, perhaps missing python libnacl package
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[DEBUG   ] LazyLoaded archive.extracted
[INFO    ] Running state [/tmp/yz/oe/oe-1.6.0.3652] at time 13:06:22.522713
[INFO    ] Executing state archive.extracted for /tmp/yz/oe/oe-1.6.0.3652
[DEBUG   ] Input seem valid so far
[DEBUG   ] Archive file /tmp/yz/oe/oe-1.6.0.3652.war is not in cache, download it
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded file.managed
[INFO    ] Running state [/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip] at time 13:06:24.455788
[INFO    ] Executing state file.managed for /var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip
[DEBUG   ] LazyLoaded config.manage_mode
[DEBUG   ] LazyLoaded file.source_list
[DEBUG   ] LazyLoaded cp.cache_file
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/salt/states/file.py", line 1545, in managed
    follow_symlinks)
  File "/usr/local/lib/python2.7/site-packages/salt/modules/file.py", line 3556, in manage_file
    dl_sum = get_hash(sfn, source_sum['hash_type'])
TypeError: string indices must be integers, not str

[ERROR   ] Unable to manage file: string indices must be integers, not str
[INFO    ] Completed state [/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip] at time 13:06:24.738531
[DEBUG   ] file.managed: {'file_|-/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip_|-/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip_|-managed': {'comment': 'Unable to manage file: string indices must be integers, not str', 'name': '/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip', 'start_time': '13:06:24.455788', 'result': False, 'duration': 282.743, '__run_num__': 0, 'changes': {}}}
[DEBUG   ] failed to download /tmp/yz/oe/oe-1.6.0.3652.war
[ERROR   ] Unable to manage file: string indices must be integers, not str
[INFO    ] Completed state [/tmp/yz/oe/oe-1.6.0.3652] at time 13:06:24.739017
[DEBUG   ] File /var/cache/salt/minion/accumulator/34555633888 does not exist, no need to cleanup.
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] LazyLoaded highstate.output
local:
----------
          ID: /tmp/yz/oe/oe-1.6.0.3652
    Function: archive.extracted
      Result: False
     Comment: Unable to manage file: string indices must be integers, not str
     Started: 13:06:22.522713
    Duration: 2216.304 ms
     Changes:

Summary
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
```

Before upgrading to salt 2015.5.3 that state worked with no errors.

**Comment by vr-jack** [comment]
Confirmed. Source using salt:// works fine, but specifying a full pathname does not.

If possible, using salt:// will act as a workaround until this is fixed; though I understand that placing a file in the saltfs may not be desired.

Also to note, md5 checking did not appear to work. Tried file and md5= using a bad md5 and no errors or warnings given. We may have to open a separate bug for that problem.

**Comment by yermulnik** [comment]
salt:// doesn't work for me:
```shell
(jail) [root#jail_salt][w:0][j:0][.../salt/states]> salt-call -l debug state.sls yz_test
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Using cached minion ID from /usr/local/etc/salt/minion_id: jail_salt
[DEBUG   ] Configuration file path: /usr/local/etc/salt/minion
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Initializing new SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Decrypting the current master AES key
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded state.sls
[DEBUG   ] LazyLoaded saltutil.is_running
[DEBUG   ] LazyLoaded grains.get
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] Fetching file from saltenv 'base', ** attempting ** 'salt://yz_test.sls'
[INFO    ] Fetching file from saltenv 'base', ** done ** 'yz_test.sls'
[DEBUG   ] LazyLoaded cmd.run
[DEBUG   ] Jinja search path: ['/var/cache/salt/minion/files/base']
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/yz_test.sls:
/tmp/yz/oe/oe-1.6.0.3652:
  archive.extracted:
    - source: /tmp/yz/oe/oe-1.6.0.3652.war
    - source_hash: salt://deploy/node_black/worker_oe/bin/oe-1.6.0.3652.war.md5
    - archive_format: zip

[DEBUG   ] LazyLoaded config.get
[DEBUG   ] Results of YAML rendering:
OrderedDict([('/tmp/yz/oe/oe-1.6.0.3652', OrderedDict([('archive.extracted', [OrderedDict([('source', '/tmp/yz/oe/oe-1.6.0.3652.war')]), OrderedDict([('source_hash', 'salt://deploy/node_black/worker_oe/bin/oe-1.6.0.3652.war.md5')]), OrderedDict([('archive_format', 'zip')])])]))])
[DEBUG   ] Error loading module.ipmi: No module named pyghmi.ipmi
[DEBUG   ] Error loading module.npm: npm execution module could not be loaded because the npm binary could not be located
[DEBUG   ] Error loading module.tls: ['PyOpenSSL version 0.14 or later must be installed before  this module can be used.']
[DEBUG   ] Error loading module.nacl: libnacl import error, perhaps missing python libnacl package
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[DEBUG   ] LazyLoaded archive.extracted
[INFO    ] Running state [/tmp/yz/oe/oe-1.6.0.3652] at time 14:22:37.002667
[INFO    ] Executing state archive.extracted for /tmp/yz/oe/oe-1.6.0.3652
[DEBUG   ] Input seem valid so far
[DEBUG   ] Archive file /tmp/yz/oe/oe-1.6.0.3652.war is not in cache, download it
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion
[DEBUG   ] Including configuration from '/usr/local/etc/salt/minion.d/_schedule.conf'
[DEBUG   ] Reading configuration from /usr/local/etc/salt/minion.d/_schedule.conf
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Loaded minion key: /usr/local/etc/salt/pki/minion/minion.pem
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[DEBUG   ] LazyLoaded file.managed
[INFO    ] Running state [/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip] at time 14:22:38.934811
[INFO    ] Executing state file.managed for /var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip
[DEBUG   ] LazyLoaded config.manage_mode
[DEBUG   ] LazyLoaded file.source_list
[DEBUG   ] LazyLoaded cp.cache_file
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/salt/states/file.py", line 1545, in managed
    follow_symlinks)
  File "/usr/local/lib/python2.7/site-packages/salt/modules/file.py", line 3556, in manage_file
    dl_sum = get_hash(sfn, source_sum['hash_type'])
TypeError: string indices must be integers, not str

[ERROR   ] Unable to manage file: string indices must be integers, not str
[INFO    ] Completed state [/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip] at time 14:22:39.222093
[DEBUG   ] file.managed: {'file_|-/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip_|-/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip_|-managed': {'comment': 'Unable to manage file: string indices must be integers, not str', 'name': '/var/cache/salt/minion/files/base/_tmp_yz_oe_oe-1.6.0.3652_.zip', 'start_time': '14:22:38.934811', 'result': False, 'duration': 287.282, '__run_num__': 0, 'changes': {}}}
[DEBUG   ] failed to download /tmp/yz/oe/oe-1.6.0.3652.war
[ERROR   ] Unable to manage file: string indices must be integers, not str
[INFO    ] Completed state [/tmp/yz/oe/oe-1.6.0.3652] at time 14:22:39.222571
[DEBUG   ] File /var/cache/salt/minion/accumulator/34555633888 does not exist, no need to cleanup.
[DEBUG   ] Re-using SAuth for ('/usr/local/etc/salt/pki/minion', 'jail_salt', 'tcp://192.168.90.19:4506')
[DEBUG   ] LazyLoaded highstate.output
local:
----------
          ID: /tmp/yz/oe/oe-1.6.0.3652
    Function: archive.extracted
      Result: False
     Comment: Unable to manage file: string indices must be integers, not str
     Started: 14:22:37.002667
    Duration: 2219.904 ms
     Changes:

Summary
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
```

**Comment by vr-jack** [comment]
it is the source that must be salt://
source_hash doesn't appear to work at all.

Also, this may be a duplicate of https://github.com/saltstack/salt/issues/25983

This state shared file.* functions so they may both be fixed with patch. Confirmation needed.

**Comment by yermulnik** [comment]
Oh, I see. Source using salt:// works fine. Thanx.

**Comment by alexdglover** [comment]
I see a potential solution to this issue was merged in #25984, does that mean the change is now 'baked' into the 2015.5.3 build for general consumption? I'm seeing the exact same issue as yermulnik, only difference is that my source_hash is being passed explicitly in the SLS file rather than referencing a salt:// md5/sha256 file

**Comment by rallytime** [comment]
@alexdglover The fix to the potential solution was added to the 2015.5 branch after 2015.5.3 was cut. However, the fix is in 2015.5.5 or newer.

@yermulnik and @vr-jack - Is this issue considered resolved? I am looking at the statements that this is likely a duplicate of #25983, which was fixed by #25984.

**Comment by alexdglover** [comment]
Thanks @rallytime !
